cl/search/docket_number_cleaner.py

```diff
@@ -0,0 +1,165 @@
+    :return: A semicolon-separated string of cleaned and normalized docket numbers in uppercase.
+    """
+    # All patterns to match
+    patterns = generic_patterns.get(Court.FEDERAL_APPELLATE, [])
+    # Combine patterns into one regex
+    combined_pattern = "|".join(patterns)
+    candidates = re.findall(combined_pattern, s, flags=re.IGNORECASE)
+    cleaned = []
+    for c in candidates:
+        # 1. Handle patterns with dash and suffix (e.g., 12-1234-ag)
+        m = re.match(
+            r"(\d{1,2})-(\d{1,6})-([a-zA-Z]{2})", c, flags=re.IGNORECASE
+        )
+        if m:
+            yy, nnnn, suffix = m.groups()
+            cleaned.append(f"{yy.zfill(2)}-{nnnn}-{suffix}")
+            continue
+        # 2. Handle patterns with dash and single-letter suffix (e.g., 12-1234P)
+        m = re.match(
+            r"(\d{1,2})-(\d{1,6})([a-zA-Z]{1})$", c, flags=re.IGNORECASE
+        )
+        if m:
+            yy, nnnn, suffix = m.groups()
+            cleaned.append(f"{yy.zfill(2)}-{nnnn}{suffix}")
+            continue
+        # 3. Handle patterns like 12-1234
+        m = re.match(r"(\d{1,2})-(\d{1,6})$", c, flags=re.IGNORECASE)
+        if m:
+            yy, nnnn = m.groups()
+            cleaned.append(f"{yy.zfill(2)}-{nnnn}")
+            continue
+        # 4. Handle patterns like 12A1234 or 12M1234
+        m = re.match(
+            r"(\d{1,2})([a-zA-Z]{1})(\d{1,6})", c, flags=re.IGNORECASE
+        )
+        if m:
+            yy, letter, nnnn = m.groups()
+            cleaned.append(f"{yy.zfill(2)}{letter}{nnnn}")
+            continue
+        # 5. Handle patterns like A-1234 or D-1234
+        m = re.match(r"([a-zA-Z]{1})-(\d{1,6})", c, flags=re.IGNORECASE)
+        if m:
+            letter, nnnn = m.groups()
+            cleaned.append(f"{letter}-{nnnn}")
+            continue
+        # 6. Handle just numbers
+        m = re.match(r"^\d{1,6}$", c, flags=re.IGNORECASE)
+        if m:
+            cleaned.append(str(int(c)))
+            continue
+
+    cleaned = [s.upper() for s in cleaned]
+    return "; ".join(cleaned)
+
+
+def clean_docket_number_raw(
+    docket_id: str, docket_number_raw: str, court_id: str
+) -> tuple[str, str]:
+    """
+    Cleans a raw docket number string based on the court's specific cleaning logic and identifies those that need LLM cleaning.
+
+    :param docket_id: The unique identifier for the docket.
+    :param docket_number_raw: The raw docket number string to be cleaned.
+    :param court_id: The identifier for the court, used to select cleaning logic.
+    :param docket_id_llm: The docket_id to be stored for downstream LLM processing if cannot be cleaned with regex. Defaults to None.
+    :return: A tuple containing the cleaned docket number and the docket_id for downstream LLM processing, if applicable.
+    """
+    court_type = court_map.get(court_id)
+
+    if court_type:
+        prelim_func = globals().get(f"prelim_clean_{court_type}")
```

Comment:
Regarding the use of globals that Semgrep suggests avoiding it if possible.

I recommend using a dispatcher method like this:

```
def get_clean_methods(court_type: str):
    match court_type.lower():
        case "f":
            return prelim_clean_F, regex_clean_F
        case _:
            return None, None
```

And call it like:
`prelim_func, regex_func = get_clean_methods(court_type)`

So this approach is also more readable and compatible with mypy.